tests/Client/UpdaterTest.php

```diff
@@ -188,69 +189,112 @@ protected function assertClientRepoVersions(array $expectedVersions): void
 
 
     /**
-     * Tests that an exception is thrown when attempting to refresh the repository with file that has an invalid valid
-     * signature.
+     * Tests that exceptions are thrown when metadata files are not valid.
      *
-     * @param string $fileToFail
-     *   The repo fail that should fail the signature check.
-     * @param integer $expectedRootVersion
-     *   The expected root version after the refresh attempt.
-     * @param string $expectionMessage
-     *   The expected exception message.
+     * @param string $fileToChange
+     *   The file to change.
+     * @param array $keys
+     *   The nested keys of the element to change.
+     * @param mixed $newValue
+     *   The new value to set.
+     * @param \Exception $expectionException
+     *   The excpected exception.
+     * @param array $expectedUpdatedVersions
+     *   The expected repo file version after refresh attempt.
      *
      * @return void
      *
-     * @dataProvider providerSignatureError
+     * @dataProvider providerRefreshException
      */
-    public function testSignatureError(string $fileToFail, int $expectedRootVersion, string $expectionMessage) : void
+    public function testRefreshException(string $fileToChange, array $keys, $newValue, \Exception $expectionException, array $expectedUpdatedVersions): void
     {
         // Use the memory storage used so tests can write without permanent
         // side-effects.
         $this->localRepo = $this->memoryStorageFromFixture('TUFTestFixtureDelegated', 'tufclient/tufrepo/metadata/current');
         $this->testRepo = new TestRepo('TUFTestFixtureDelegated');
-        $this->assertClientRepoVersions(static::getFixtureClientStartVersions('TUFTestFixtureDelegated'));
-        $this->testRepo->setRepoFileNestedValue($fileToFail);
+        $this->assertClientRepoVersions($this->getFixtureClientStartVersions('TUFTestFixtureDelegated'));
+        $this->testRepo->setRepoFileNestedValue($fileToChange, $keys, $newValue);
         $updater = $this->getSystemInTest();
         try {
             $updater->refresh();
-        } catch (SignatureThresholdExpception $exception) {
-            // Confirm the root was updated to version 4 but not to 5 because
-            // 5.root.json should throw an exception.
-            $this->assertSame(
-                $expectedRootVersion,
-                RootMetadata::createFromJson($this->localRepo['root.json'])->getVersion()
-            );
-            $this->assertSame($expectionMessage, $exception->getMessage());
+        } catch (\Exception $exception) {
+            $this->assertEquals($exception, $expectionException);
+            $this->assertClientRepoVersions($expectedUpdatedVersions);
             return;
         }
-        $this->fail('No SignatureThresholdExpception thrown');
+        $this->fail('No exception thrown. Expected: ' . get_class($expectionException));
```

Comment:
Can't we use `$this->expectException()` instead?

Reply:
I don't think so because we want to check `$this->assertClientRepoVersions($expectedUpdatedVersions);` after the exception is thrown. 

For instance for the root metadata object is updated because the new one is valid that change should persist even if we get an exception on the snapshot. But the snapshot should not persist. So need to check the version after the exception